File: web/blog/media/prototype_one/data/field.js
data.push( { type : 'file', name : 'field.h', content : [





	{ line : 7, type : 'class', name : 'Field', content : {
		public : [
			{ line : 9, type : 'class', name : 'Token', content : { public : [
					{ line : 10, type : 'member', name : 'None' },
					{ line : 11, type : 'member', name : 'PlayerA' },
					{ line : 12, type : 'member', name : 'PlayerB' }
			]}},

			{ line : 15, type : 'method', name : 'Opponent', abstraction : 'static', return : 'Token', params : [ 'Token' ], content : [
				{ line : 16, type : 'call', function : 'assert' }, { line : 16, type : 'use', class : 'Token', member : 'None' },
				{ line : 17, type : 'use', class : 'Token', member : 'PlayerA' },
				{ line : 18, type : 'use', class : 'Token', member : 'PlayerB' },

				{ line : 20, type : 'use', class : 'Token', member : 'PlayerA' }

			]},

			{ line : 24, type : 'class', name : 'Move', content : { public : [
					{ line : 25, type : 'member', name : 'row', class : 'int' },
					{ line : 26, type : 'member', name : 'col', class : 'int' }
			]}},

			{ line : 29, type : 'method', name : 'Field', content : [
				{ line : 30, type : 'use', member : 'left_' },
				{ line : 31, type : 'use', member : 'grid_' }, { line : 31, type : 'use', class : 'Token' },

				{ line : 33, type : 'use', member : 'grid_' }, { line : 33, type : 'use', class : 'Token' },

			]},

			{ line : 37, type : 'method', name : '~Field', content : [

				{ line : 39, type : 'use', member : 'grid_' },

				{ line : 41, type : 'use', member : 'grid_' },
			]},

			{ line : 44, type : 'method', name : 'Clone', return : 'Field', content : [



				{ line : 48, type : 'use', member : 'grid_' }, { line : 48, type : 'use', member : 'grid_' },


				{ line : 51, type : 'use', member : 'left_' }, { line : 51, type : 'use', member : 'left_' }
			]},


			{ line : 55, type : 'method', name : 'Clear', content : [


				{ line : 58, type : 'use', member : 'grid_' }, { line : 58, type : 'use', class : 'Token', member : 'None' },


				{ line : 61, type : 'use', member : 'left_' }
			]},

			{ line : 64, type : 'method', name : 'Show', content : [
				{ line : 65, type : 'use', class : 'std::cout' },

				{ line : 67, type : 'use', class : 'std::cout' },


				{ line : 70, type : 'use', member : 'grid_' }, { line : 70, type : 'use', class : 'Token', member : 'PlayerA' },
				{ line : 71, type : 'use', class : 'std::cout' },
				{ line : 72, type : 'use', member : 'grid_' }, { line : 71, type : 'use', class : 'Token', member : 'PlayerA' },
				{ line : 73, type : 'use', class : 'std::cout' },

				{ line : 75, type : 'use', class : 'std::cout' },



				{ line : 79, type : 'use', class : 'std::cout' },




				{ line : 84, type : 'use', class : 'std::cout' },


				{ line : 87, type : 'use', class : 'std::cout' }
			]},

			{ line : 90, type : 'method', name : 'SameInRow', return : 'int', params : [ 'Token', 'int' ], content : [




				{ line : 95, type : 'use', member : 'grid_' }, { line : 95, type : 'use', member : 'grid_' }, { line : 95, type : 'use', member : 'grid_' },

				{ line : 97, type : 'use', member : 'grid_' }, { line : 97, type : 'use', member : 'grid_' }, { line : 97, type : 'use', member : 'grid_' },




				{ line : 102, type : 'use', member : 'grid_' }, { line : 102, type : 'use', member : 'grid_' }, { line : 102, type : 'use', member : 'grid_' },

				{ line : 104, type : 'use', member : 'grid_' }, { line : 104, type : 'use', member : 'grid_' }, { line : 104, type : 'use', member : 'grid_' },




			]},

			{ line : 111, type : 'method', name : 'InRange', return : 'bool', params : [ 'Move' ], content : [
				{ line : 112, type : 'use', class : 'Move', member : 'row' }, { line : 112, type : 'use', class : 'Move', member : 'row' }, { line : 112, type : 'use', class : 'Move', member : 'col' }, { line : 112, type : 'use', class : 'Move', member : 'col' }
			]},

			{ line : 115, type : 'method', name : 'IsEmpty', return : 'bool', params : [ 'Move' ], content : [
				{ line : 116, type : 'use', member : 'grid_' }, { line : 116, type : 'use', class : 'Move', member : 'row' }, { line : 116, type : 'use', class : 'Move', member : 'col' }, { line : 116, type : 'use', class : 'Token', member : 'None' }
			]},

			{ line : 119, type : 'method', name : 'IsFull', return : 'bool', content : [
				{ line : 120, type : 'use', member : 'left_' }
			]},

			{ line : 123, type : 'method', name : 'MakeMove', params : [ 'Move', 'Token' ], content : [
				{ line : 124, type : 'call', function : 'assert' }, { line : 124, type : 'call', method : 'InRange' },
				{ line : 125, type : 'call', function : 'assert' }, { line : 125, type : 'call', method : 'IsEmpty' },
				{ line : 126, type : 'call', function : 'assert' }, { line : 126, type : 'use', class : 'Token', member : 'None' },
				{ line : 127, type : 'call', function : 'assert' }, { line : 127, type : 'use', member : 'left_' },

				{ line : 129, type : 'use', member : 'grid_' }, { line : 129, type : 'use', class : 'Move', member : 'row' }, { line : 129, type : 'use', class : 'Move', member : 'row' },
				{ line : 130, type : 'use', member : 'left_' },
			]},

			{ line : 133, type : 'method', name : 'ClearMove', params : [ 'Move' ], content : [
				{ line : 134, type : 'call', function : 'assert' }, { line : 134, type : 'call', method : 'InRange' },
				{ line : 135, type : 'call', function : 'assert' }, { line : 135, type : 'call', method : 'IsEmpty' },
				{ line : 136, type : 'call', function : 'assert' }, { line : 136, type : 'use', member : 'left_' },

				{ line : 138, type : 'use', member : 'grid_' }, { line : 138, type : 'use', class : 'Move', member : 'row' }, { line : 138, type : 'use', class : 'Move', member : 'row' }, { line : 138, type : 'use', class : 'Token', member : 'None' },
				{ line : 139, type : 'use', member : 'left_' },
			]}

		], private : [
			{ line : 143, type : 'member', name : 'grid_', class : 'Token' },
			{ line : 144, type : 'member', name : 'left_', class : 'int' }
		]}
	}
]});

File: bin/app/user/projects/tictactoe_cpp/src/field.cpp
#include "field.h"

#include "io.h"

Field::Token Field::Opponent( Token token ) {
	if ( token == PlayerA ) {
		return PlayerB;
	} else if (token == PlayerB){
		return PlayerA;
	} else {
		return None;
	}
}

Field::Field()
	: left_( 9 ) {
	grid_ = new Token* [3];
	for ( int i = 0; i < 3 ; i++ ) {
		grid_[i] = new Token[3];
	}
}

Field::~Field() {
	for ( int i = 0; i < 3; i++ ) {
		delete [] grid_[i];
	}
	delete [] grid_;
}

Field Field::Clone() const {
	Field field;
	for ( int i = 0; i < 3; i++ ) {
		for ( int j = 0; j < 3; j++ ) {
			field.grid_[i][j] = grid_[i][j];
		}
	}
	field.left_ = left_;
	return field;
}

void Field::Clear() {
	for ( int i = 0; i < 3; i++ ) {
		for ( int j = 0; j < 3; j++ ) {
			grid_[i][j] = None;
		}
	}
	left_ = 9;
}

void Field::Show() const {
	io::stringOut("   1   2   3\n");
	for ( int row = 0; row < 3; row++ ) {
		io::numberOut(row + 1);
		io::stringOut(" ");

		for ( int col = 0; col < 3; col++ ) {
			if ( grid_[row][col] == PlayerA ) {
				io::stringOut(" X ");
			} else if ( grid_[row][col] == PlayerB ) {
				io::stringOut(" O ");
			} else {
				io::stringOut("   ");
			}

			if ( col < 2 ) {
				io::stringOut("|");
			}
		}

		if ( row < 2 ) {
			io::stringOut("\n  -----------\n");
		}
	}
	io::stringOut("\n\n");
}

int Field::SameInRow( Token token, int amount ) const {
	int sum = amount * token;
	int count = 0;

	for ( int i = 0; i < 3; i++ ) {
		if ( grid_[i][0] + grid_[i][1] + grid_[i][2] == sum ) {
			count++;
		} 
		if ( grid_[0][i] + grid_[1][i] + grid_[2][i] == sum ) {
			count++;
		}
	}

	if ( grid_[0][0] + grid_[1][1] + grid_[2][2] == sum ) {
		count++;
	} 
	if ( grid_[2][0] + grid_[1][1] + grid_[0][2] == sum ) {
		count++;
	}

	return count;
}

bool Field::InRange( const Move& move ) const {
	return move.row >= 0 && move.row < 3 && move.col >= 0 && move.col < 3;
}

bool Field::IsEmpty( const Move& move ) const {
	return grid_[move.row][move.col] == None;
}

bool Field::IsFull() const {
	return left_ == 0;
}

void Field::MakeMove( const Move& move, Token token ) {
	if ( !InRange( move ) || !IsEmpty( move ) || token == None || !left_ ) {
		return;
	}

	grid_[move.row][move.col] = token;
	left_--;
}

void Field::ClearMove( const Move& move ) {
	if ( !InRange( move ) || IsEmpty( move ) || left_ == 9 ) {
		return;
	}

	grid_[move.row][move.col] = None;
	left_++;
}
